### PR TITLE
Added meshkit style errors and comments

### LINKED ISSUE #4262: Need a probable cause and suggested remediation for errors in k8s component generation
Will need a probable cause and suggested remediation for these errors.

_Originally posted by @leecalcote in https://github.com/meshery/meshery/pull/4254#discussion_r717029551_

### PR DESCRIPTION
**Pull request by Revolyssup**
Signed-off-by: ashish <ashishjaitiwari15112000@gmail.com>

**Description**

This PR fixes #4262 

**Notes for Reviewers**

**[Signed commits](https://github.com/meshery/meshery/blob/master/CONTRIBUTING.md#signing-off-on-commits-developer-certificate-of-origin)**
- [x] Yes, I signed my commits.

### CODE REVIEW

File: models/pattern/core/error.go [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+import (
+	"github.com/layer5io/meshkit/errors"
+)
+
+const (
+	ErrGetK8sComponentsCode = "10000"
+)
+
+func ErrGetK8sComponents(err error) error {
+	return errors.New(ErrGetK8sComponentsCode, errors.Alert, []string{"Could not get K8s components for registeration"}, []string{err.Error()}, []string{"Got an invalid kubeconfig", "Filters passed incorrectly in config", "Could not fetch api resources from kubernetes server"}, []string{"Make sure that the configuration filters passed are in accordance with output from /openapi/v2"})
```
Comment (leecalcote) [inline models/pattern/core/error.go:12]:
```suggestion
	return errors.New(ErrGetK8sComponentsCode, errors.Alert, []string{"Could not get K8s components for registration"}, []string{err.Error()}, []string{"Invalid kubeconfig", "Filters passed incorrectly in config", "Could not fetch API resources from Kubernetes server"}, []string{"Make sure that the configuration filters passed are in accordance with output from /openapi/v2"})
```

File: models/pattern/core/error.go [hunk 2]
```diff
@@ -0,0 +1,13 @@
+package core
+
+import (
+	"github.com/layer5io/meshkit/errors"
+)
+
+const (
+	ErrGetK8sComponentsCode = "10000"
```
Comment (leecalcote) [inline models/pattern/core/error.go:8]: Should this be statically set or should the error code utility determine the appropriate #?, @DelusionalOptimist

Comment (leecalcote) [reply models/pattern/core/error.go:8]: @Revolyssup fyi - https://docs.meshery.io/project/contributing-error.... "replace_me" instead?

Comment (Revolyssup) [reply models/pattern/core/error.go:8]: will make the change

**Comment by Revolyssup** [comment]
@leecalcote have changed the filter according to kubeopenapi changes also

File: handlers/k8sconfig_handler.go [hunk 3]
```diff
@@ -234,7 +234,12 @@ func (h *Handler) checkIfK8SConfigExistsOrElseLoadFromDiskOrK8S(req *http.Reques
 		}
 		prefObj.K8SConfig = kc
 		ctxID := "0" // To be replaced after multi-context support
-		go registerK8sComponents(h.log, prefObj.K8SConfig.Config, ctxID)
+		go func(l logger.Handler, config []byte, ctx string) {
+			err := registerK8sComponents(h.log, prefObj.K8SConfig.Config, ctxID)
+			if err != nil {
+				logrus.Error(err)
```
Comment (leecalcote) [inline handlers/k8sconfig_handler.go:240]: switch to MeshKit errors, please.

**Review by leecalcote — APPROVED** [review]

File: handlers/k8sconfig_handler.go (hunk 3 above)
Comment (Revolyssup) [reply handlers/k8sconfig_handler.go:240]: the error returned by registerK8sComponents are meshkit errors only

**Review by DelusionalOptimist — APPROVED** [review]
This looks good. :rocket: 
However,
```
time="2021-09-28T19:20:37Z" level=error msg="open /home/appuser/.meshery/bin/kubeopenapi-jsonschema: no such file or directory"
```
* Need to pack jsonschema utility in Meshery's docker image
* In case it doesn't exist, it should be downloaded on the fly but I couldn't find it at the target location in my container.

**Comment by Revolyssup** [comment]
> This looks good. 🚀 However,
> 
> ```
> time="2021-09-28T19:20:37Z" level=error msg="open /home/appuser/.meshery/bin/kubeopenapi-jsonschema: no such file or directory"
> ```
> 
> * Need to pack jsonschema utility in Meshery's docker image
> * In case it doesn't exist, it should be downloaded on the fly but I couldn't find it at the target location in my container.
We don't need to package it in meshery. 
It is being fetched on the fly. 
The error you are getting is because bin directory is not present in container. can you add that in docker image. Download will happen correctly then.

**Comment by leecalcote** [comment]
@Revolyssup .... the binary for this utility should be packaged into the container.

**Comment by Revolyssup** [comment]
> @Revolyssup .... the binary for this utility should be packaged into the container image.

Okay will add

**Comment by DelusionalOptimist** [comment]
> The error you are getting is because bin directory is not present in container. can you add that in docker image. Download will happen correctly then.

Hmm... I think the function (in meshkit) responsible for fetching this should take care of it. When we used fetched service-mesh CLIs that directory did get created :thinking:. Anyways prolly something to do later.

**Comment by DelusionalOptimist** [comment]
I hope glibc has all the shared libs needed by this binary :crossed_fingers:

**Comment by DelusionalOptimist** [comment]
@Revolyssup I was able to build a container using this image and successfully run the jsonschema utility in it. It will work fine.
The function is also detecting the binary at the location and not fetching it.
However, I see this in the logs. Do look into what exactly is giving this error. Hopefully it isn't bad arguments given to the utility :crossed_fingers: 
```
time="2021-09-28T20:42:33Z" level=error msg="exit status 1"
```

**Comment by leecalcote** [comment]
The current stable release crashes on me with this bug.

File: internal/store/store.go [hunk 4]
```diff
@@ -46,6 +46,8 @@ func Set(key string, value interface{}) {
 
 	globalStore.store[key][hash] = value
 }
+
+// Delete will take the key and values which needs to be deleted from the global store and delete that entry
```
Comment (leecalcote) [inline internal/store/store.go:50]: Thank you, @Revolyssup 🥇

**Comment by leecalcote** [comment]
Gents, my 🤞 that this patches the stable release and that the binary failure is addressed shortly. @utkarsh-pro is this due to the latest release of that tool?

**Review by leecalcote — APPROVED** [review]

**Comment by utkarsh-pro** [comment]
> Gents, my crossed_fingers that this patches the stable release and that the binary failure is addressed shortly. @utkarsh-pro is this due to the latest release of that tool?

I don't think the newer release is actually being used anywhere at the moment. :thinking:

**Comment by Revolyssup** [comment]
> > Gents, my crossed_fingers that this patches the stable release and that the binary failure is addressed shortly. @utkarsh-pro is this due to the latest release of that tool?
> 
> I don't think the newer release is actually being used anywhere at the moment. thinking

I have just opened the PR for version upgrade in meshkit. And currently the kubep[embinary that is seeded is of an older version, so that too would be required to updated

**Comment by DelusionalOptimist** [comment]
All, @Revolyssup pointed out that moving to new release of the jsonchema utility fixes the above bug.
I've verified it by running the image with latest version of the utility and Meshery in a container. Its working. Also, k8s native components are being registered successfully :+1: 

See #4272

**Pull request merged**